### ISSUE TITLE
Correct usage of post_camera_hook_fn?

### ISSUE THREAD

**Issue by dpshepherd** [micro-manager/pycro-manager#160]
Hi all,

We are porting a functioning Micromanager 2.0 beanshell script to run a custom microscope. We need to use the `post_camera_hook_fn` feature for this control code to work correctly.

If we create a bridge to Micromanager, setup the camera to wait for a trigger, setup our stage controller to provide the trigger, start the camera in micromanager itself, and then execute the hook function, everything works as expected.

However, when we try to use the `Acquisition` class, we do not get the expected behavior. The acquisition opens, but the stage does not provide the camera a trigger or start moving. No frames are acquired in the acquisition. This indicates to us that the `post_camera_hook_fn` function is not executing. We thought it might be a timeout on the camera waiting for a trigger, but we set that to 60 s and this issue still occurs.

The notebook to control the microscope is [here](https://github.com/QI2lab/OPM/blob/master/Control-MM/OPM_control_noTeensy.ipynb).

Any suggestions?

Thanks!

**Comment by henrypinkard** [comment]
Just to make sure, youre using latest versions of micro-mangaer/pycro-manager, right?

Need to figure out if the hook is not executing or if its problem with the hardware. I run through PyCharm usually instead of a notebook. If you copy and paste your code in there, you can set a breakpoint within the body of the hook function, see if it triggers, and manually type your command to the hardware in there. I just checked and it does trigger with the demo camera.

**Comment by dpshepherd** [comment]
`conda list ` gives:
`pycromanager              0.5.0                    pypi_0    pypi`

and Micromanager is nightly build 20201014.

We put a `print` statement in the hook function in the notebook, which outputs when calling outside of `Acquisition` and does not call when we try to use `post_camera_hook_fn`. 

One suspicion is the camera driver. We have had and continue to have a ton of issues with the Photometrics drivers. The stage is fairly simple and just takes serial commands.

**Comment by dpshepherd** [comment]
Ok, so when I put into VS code, I pylint gives me a warning that `post_camera_hook_fn` is unexpected. If I look into acquire.py on our machine at the definition of `Acquisition`, the class definition is:

```
class Acquisition(object):
    def __init__(self, directory=None, name=None, image_process_fn=None,
                 pre_hardware_hook_fn=None, post_hardware_hook_fn=None, show_display=True, tile_overlap=None,
                 magellan_acq_index=None, process=True, debug=False):
```

This different from acquire.py that is in the repo here, which lists:

```
class Acquisition(object):
    """ """
    def __init__(self, directory=None, name=None, image_process_fn=None,
                 pre_hardware_hook_fn=None, post_hardware_hook_fn=None, post_camera_hook_fn=None,
                 show_display=True, tile_overlap=None, max_multi_res_index=None,
                 magellan_acq_index=None, magellan_explore=False, process=False, debug=False):
```

I'll pull the repo separately and try again.

**Comment by dpshepherd** [comment]
Ok...there was an old installation of pycromanager on the machine in a different environment, I wiped things clean, pulled the GitHub repo, and started fresh. Now the acquire.py matches the online repo.

I set a breakpoint in the hook function and it never reaches it.

**Comment by henrypinkard** [comment]
What about if you run it with micromanahger demo config? Does it reach the breakpoint in the hook function?

**Comment by henrypinkard** [comment]
you can run this: https://github.com/henrypinkard/pycro-manager/blob/master/test/acq_hook.py

just change it to a `post_camera_hook_fn`

**Comment by henrypinkard** [comment]
also might be worth checking micromanager core log to see if any errors show up there

**Comment by dpshepherd** [comment]
What devices do you have in the demo config?

**Comment by henrypinkard** [comment]
just the normal ones you that come with the config file you get when you download a nightly build. Theres a bunch of them but all you need is the camera

**Comment by dpshepherd** [comment]
All the core log has is:

```
2020-10-16T08:50:48.100692 tid2632 [IFO,App] Finished updating GUI
2020-10-16T08:51:45.813526 tid10632 [IFO,dev:Coherent-Scientific Remote] requested setpoint: 1 but echo setpoint is: 1
2020-10-16T08:52:11.824801 tid2632 [IFO,App] Refreshing config group table
2020-10-16T08:52:12.005697 tid2632 [IFO,App] Finished refreshing config group table
2020-10-16T08:52:12.005697 tid2632 [IFO,App] Updating GUI; config pad = false; from cache = false
2020-10-16T08:52:12.564625 tid2632 [IFO,App] Finished updating GUI
```

**Comment by dpshepherd** [comment]
With the demo configuration, I hit the breakpoint in `hook_fn`. Once I step forward, the acquisition display appears. But, no images appear.

If I remove the breakpoint, the acquisition display appears but no images are populated. 

If I hit "live" in Micromanger, I get the stripes one expects from the demo camera.

**Comment by dpshepherd** [comment]
If I shutdown Micromanager and restart it, I can get the demo camera to return images when using `post_camera_hook_fn`.

**Comment by dpshepherd** [comment]
I was able to get our stage moving by restarting the computer and reinstalling Micromanager. This indicates to me that if we send an incorrect command, we lock up Micromanager.

There seem to be other problems now and we get the following error when trying to talk to the stage controller after the acquisition finishes:
```
Traceback (most recent call last):
  File "c:/Users/dpsheph1/Documents/GitHub/OPM/Control-MM/test_pycrom.py", line 381, in <module>
    main()
  File "c:/Users/dpsheph1/Documents/GitHub/OPM/Control-MM/test_pycrom.py", line 368, in main
    core.set_property('CameraTrigger','Internal Trigger')
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 290, in <lambda>
    fn = lambda instance, *args, signatures_list=tuple(methods_with_name): instance._translate_call(signatures_list, args)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 373, in _translate_call
    valid_method_spec, deserialize_types = _check_method_args(method_specs, fn_args)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 523, in _check_method_args
    for java_arg_class, python_arg_val in zip(valid_method_spec['arguments'], fn_args):
TypeError: 'NoneType' object is not subscriptable
Exception ignored in: <function JavaObjectShadow.__del__ at 0x000001E1D31531F0>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
```

**Comment by henrypinkard** [comment]
That sounds right. I just made a change that will give you a more interpretable error message on that last one (https://github.com/micro-manager/pycro-manager/pull/161).

Just to confirm, if you start up everything fresh, use only the demo configuration, and use the `post_camera_hook_fn`, the sine wave pattern still appears, correct?

And then the problems come when your particular hardware gets stuck in the wrong state?

**Comment by dpshepherd** [comment]
Now, we are able to get the hardware to execute one scan each time after wiping and reinstalling Micromanager. The file writing seems to be a bottleneck at this point. The SSD RAID array is not maxing out (writing at ~60 MB/s, capable of ~600 MB/s), but the disk is active long after the stage scan completes.

The buffer in Micromanager does not show any usage (0/6000 images filled).  Is there another way to diagnose how many images have been acquired vs written?

If I use the MDA acquisition window, I can write data at a x10 faster framerate without using the buffer. Past that, we start piling images up in the buffer.

**Comment by dpshepherd** [comment]
I caught the problem that was causing the above exception, but now get the following when calling core again after the acquisition,

```
Exception ignored in: <function JavaObjectShadow.__del__ at 0x000001FE18C841F0>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
```

This is the current file: https://github.com/QI2lab/OPM/blob/master/Control-MM/test_pycrom.py

**Comment by henrypinkard** [comment]
> The demo config is working fine. If I declare the wrong number or wrong order of arguments when defining hook_fn, I can reproduce the issue and have to restart Micromanager.
 
This is not surprising to me. Might be more trouble than its worth to fix at the moment

> Now, we are able to repeatedly get the hardware to execute one scan each time. However, the acquisition is active after the stage finishes the scan and the disk stays active for a minute or so after the scan. However, the SSD RAID array is not maxing out (writing at ~60 MB/s, capable of ~600 MB/s).
> 
> The buffer in Micromanager does not show any usage (0/6000 images filled). Is there another way to diagnose how many images have been acquired vs written?

I suspect this is because you're doing one acquisition for each time point, rather than a single acquisition that you repeatedly send acquire commands to. Ideally, you would set up all that stuff that is happening in functions surrounding where `Acquisition` is now to run in acquisition hooks, then you add all channels/tiles/other axes into your events, and do all your hardware setup in the hook functions as appropriate

> If I use the MDA acquisition window to acquire a timelapse with the interval = 0 ms, the system can write data at a x10 faster framerate without using the buffer. Past that, images pile up in the buffer.
> 
> This behavior is different from older version of Micromanager using our Beanshell script.

What do you mean, the speed of the micro-manager MDA?

**Comment by henrypinkard** [comment]
> I caught the problem that was causing the above exception, but now get the following when calling core again after the acquisition,

Hard for me to tell what is causing this with the way code is setup now, maybe fix the above and it will be more clear

**Comment by dpshepherd** [comment]
> This is not surprising to me. Might be more trouble than its worth to fix at the moment

Not suggesting that you do. Just documenting what we did wrong in case others run into this.

> I suspect this is because you're doing one acquisition for each time point, rather than a single acquisition that you repeatedly send acquire commands to. Ideally, you would set up all that stuff that is happening in functions surrounding where Acquisition is now to run in acquisition hooks, then you add all channels/tiles/other axes into your events, and do all your hardware setup in the hook functions as appropriate

I really don't understand how this would make a difference in the test case of M=1 strips. We aren't acquiring time lapse data. We are executing a large area scan by moving the X stage at a constant speed while the camera acquires at a constant rate. Each "strip" is one laser, one y axis position, one z stage position, and N oblique light sheet images. N is usually something like 1-5 million images in our real experiments. 

Here, we are using N=5000 images in M=1 strip for testing. So there is no repeated call to the `Acquisition `class. This indicates that rewriting how we are controlling our hardware would not fix this problem.

> What do you mean, the speed of the micro-manager MDA?

The speed of the acquisition when writing to disk. If we setup a timelapse series and acquire say 10,000 frames with interval = 0 ms, the SSD raid array and Micromanager can keep up writing to disk so that there are no images held in buffer. This is also true if we use our Beanshell script instead of Pycromanager.

**Comment by henrypinkard** [comment]
> Here, we are using N=5000 images in M=1 strip for testing. So there is no repeated call to the `Acquisition `class. This indicates that rewriting how we are controlling our hardware would not fix this problem.

Ah okay didn't realize it was only running once

**Comment by dpshepherd** [comment]
> Hard for me to tell what is causing this with the way code is setup now, maybe fix the above and it will be more clear

It is not clear to me that figuring out how to rewrite everything into your specific hooks, when we cannot get the M=1 example to work is a "fix". I just stripped out all of the hardware calls you were worried about and still get this behavior.

**Comment by henrypinkard** [comment]
Not sure why it would be slower. @ieivanov and I were able to optimize it so that it was as fast or faster than the micromanager MDA. @ieivanov is it still going at that speed on your system?

**Comment by dpshepherd** [comment]
Assuming we will figure out the writing, speed, we are still a bit confused about how to use the event structure for our particular type of imaging that relies on the `post_camera_hook_fn`.

Right now we have two issues:

### First issue

We tested that we can move the stage scanning setup code into a function and call that using `pre_hardware_hook_fn`. That way, we will always setup the stage scan during the `acq.acquire(events)` call. When we do this the scan starts, but the acquisition does not return back to Python. The code hangs and we have to kill it. I'm assuming we are doing something incorrectly with declaring events or the `pre_hardware_hook_fn` itself.

That code is here: [test_stripped_down_pycrom.py](https://github.com/QI2lab/OPM/blob/master/Control-MM/test_stripped_down_pycrom.py)

### Second issue 

Beyond that, we are still not totally clear on how to create the events for our acquisitions. For example. If we have two position of the tiling stage [y1, y2] and N=100 images we want to acquire in our constant stage scan, we could create the following structure:
```
number_of_images=10
tile_axis_positions_asi=[150000,152000]
events = []
for y in range(len(tile_axis_positions_asi)):
    for z in range(number_of_images-1):
        events.append({'axes': {'tile': y, 'scan': z}, 'y': tile_axis_positions_asi[y]})
```

This gives,
[{'axes': {'tile': 0, 'scan': 0}, 'y': 150000},
 {'axes': {'tile': 0, 'scan': 1}, 'y': 150000},
 {'axes': {'tile': 0, 'scan': 2}, 'y': 150000},
 {'axes': {'tile': 0, 'scan': 3}, 'y': 150000},
 {'axes': {'tile': 0, 'scan': 4}, 'y': 150000},
 {'axes': {'tile': 0, 'scan': 5}, 'y': 150000},
 {'axes': {'tile': 0, 'scan': 6}, 'y': 150000},
 {'axes': {'tile': 0, 'scan': 7}, 'y': 150000},
 {'axes': {'tile': 0, 'scan': 8}, 'y': 150000},
 {'axes': {'tile': 0, 'scan': 9}, 'y': 150000},
 {'axes': {'tile': 1, 'scan': 0}, 'y': 152000},
 {'axes': {'tile': 1, 'scan': 1}, 'y': 152000},
 {'axes': {'tile': 1, 'scan': 2}, 'y': 152000},
 {'axes': {'tile': 1, 'scan': 3}, 'y': 152000},
 {'axes': {'tile': 1, 'scan': 4}, 'y': 152000},
 {'axes': {'tile': 1, 'scan': 5}, 'y': 152000},
 {'axes': {'tile': 1, 'scan': 6}, 'y': 152000},
 {'axes': {'tile': 1, 'scan': 7}, 'y': 152000},
 {'axes': {'tile': 1, 'scan': 8}, 'y': 152000},
 {'axes': {'tile': 1, 'scan': 9}, 'y': 152000}]``

How do we ensure that the constant speed stage scan that the hardware triggering is placed into the 'scan' event **and** that Pycromanager will not attempt to talk to the y-axis stage during 10 images that are acquired in the hardware controlled constant speed stage scan? I think if we can solve the issue that we cannot send commands to stage controller during the constant speed scan, then we can structure our events to use hooks. If not, we will need to use the loops that we used in the Beanshell script.

Thanks!

**Comment by ieivanov** [comment]
> Not sure why it would be slower. @ieivanov and I were able to optimize it so that it was as fast or faster than the micromanager MDA. @ieivanov is it still going at that speed on your system?

I'll try to test again on Monday. Does the `post_camera_hook_fn` shuttle image data from java into python (like `image_process_fn` does)? That part will still be slow. If the data stays on the java side, I'd expect the acquisition to go quickly.

**Comment by dpshepherd** [comment]
Reading through the API some more, we are starting to see a potential path forward to using the events, assuming we can synchronize everything. The ASI Tiger controller is quite picky about timing and communications.

If we place a `print(event)` call inside of the functions we call for `post_hardware_hook_fn` and `post_camera_hook_fn`, we see the following, in this order:

1. the full list/dict due to a call to `post_hardware_hook_fn` 
2. the full list/dict due to a call to `post_camera_hook_fn`
3. an empty list/dict due to a call to `post_camera_hook_fn`

I don't quite understand this. I thought these functions would just get the current event? Also, why is the camera called again at the end?

Thanks again, I know we have a complicated use case.

**Comment by henrypinkard** [comment]
> I'll try to test again on Monday. Does the `post_camera_hook_fn` shuttle image data from java into python (like `image_process_fn` does)? That part will still be slow. If the data stays on the java side, I'd expect the acquisition to go quickly.

Thanks! No, it should only be tiny messages going back and forth, plus `post_camera_hook_fn` only runs once for each sequence. Just wondering if the fast writing still works on your system with all the other features stripped away, so that we can tell if I broke something or if its something specific to @dpshepherd 's configuration

**Comment by henrypinkard** [comment]
> We tested that we can move the stage scanning setup code into a function and call that using `post_hardware_hook_fn`. That way, we will always setup the stage scan during the `acq.acquire(events)` call. When we do this the scan starts, but the acquisition does not return back to Python. The code hangs and we have to kill it. I'm assuming we are doing something incorrectly with declaring events or the `post_hardware_hook_fn` itself.
> 
> That code is here: [test_stripped_down_pycrom.py](https://github.com/QI2lab/OPM/blob/master/Control-MM/test_stripped_down_pycrom.py)

Code looks right to me. Can you strip away hardware calls and reproduce with demo configuration?

**Comment by henrypinkard** [comment]
> How do we ensure that the images acquired during the constant speed stage scan are placed into the 'scan' event axis **and** that Pycromanager will not attempt to talk to the y-axis stage during 10 images that are acquired in the hardware controlled constant speed stage scan?

I think the acquisition engine won't try to set a new y position if it doesn't change from event to event. Are you seeing something different? Maybe some of that logic doesn't work because you just have Y, not X. Or did you just take away X in that example for clarity?

**Comment by dpshepherd** [comment]
> > We tested that we can move the stage scanning setup code into a function and call that using `post_hardware_hook_fn`. That way, we will always setup the stage scan during the `acq.acquire(events)` call. When we do this the scan starts, but the acquisition does not return back to Python. The code hangs and we have to kill it. I'm assuming we are doing something incorrectly with declaring events or the `post_hardware_hook_fn` itself.
> > That code is here: [test_stripped_down_pycrom.py](https://github.com/QI2lab/OPM/blob/master/Control-MM/test_stripped_down_pycrom.py)
> 
> Code looks right to me. Can you strip away hardware calls and reproduce with demo configuration?

Yes, this behavior still occurs with the demo config. And we see the same behavior with `print(event)` as above.

**Comment by dpshepherd** [comment]
> > How do we ensure that the images acquired during the constant speed stage scan are placed into the 'scan' event axis **and** that Pycromanager will not attempt to talk to the y-axis stage during 10 images that are acquired in the hardware controlled constant speed stage scan?
> 
> I think the acquisition engine won't try to set a new y position if it doesn't change from event to event. Are you seeing something different? Maybe some of that logic doesn't work because you just have Y, not X. Or did you just take away X in that example for clarity?

Removed for clarity. Right now, the path forward seems to be parsing the event structure in the hook functions to configure channels, move the stage in the Y, or move the stage in Z when the X stage is finished scanning. If the ASI Tiger controller receives a command during the constant speed stage scan, it stops scanning. So we have to avoid this situation.

Beyond stage communication, we normally use a homebuilt DAQ setup to control the lasers. So we'll also need to be able to send serial commands to it when the channel changes, instead of the standard channel definition in Micromanager. The laser controller is also not expecting communication during the constant speed stage scan.

**Comment by henrypinkard** [comment]
I think the XY stage would be fine with the automatic control, but if you're putting stuff into the hooks anyway control hardware that should work  too

I was able to track down the bug with the post camera hook. Post camera hooks were inadvertently running as post hardware hooks, and not shutting down properly. Yikes! So this should explain several of the weird behaviors you saw. Thanks for the detective work! It is merging into nihgtly builds after this PR (https://github.com/micro-manager/micro-manager/pull/984). Or if you want to use it now, drop in the latest Jar from [here](https://dl.bintray.com/micro-manager/micro-manager/org/micro-manager/acqengj/AcqEngJ/0.7.2/) in the plugins/micro-manager folder of your micromanaegr install

**Comment by dpshepherd** [comment]
> I think the XY stage would be fine with the automatic control, but if you're putting stuff into the hooks anyway control hardware that should work too

I think the issue is that the X stage is the scanning stage. So any normal update calls can interrupt the constant speed scan. This problem is why we arrived at such an awkward Beanshell script. We'll try both and see.

> I was able to track down the bug with the post camera hook. Post camera hooks were inadvertently running as post hardware hooks, and not shutting down properly. Yikes! So this should explain several of the weird behaviors you saw. Thanks for the detective work! It is merging into nihgtly builds after this PR ([micro-manager/micro-manager#984](https://github.com/micro-manager/micro-manager/pull/984)). Or if you want to use it now, drop in the latest Jar from [here](https://dl.bintray.com/micro-manager/micro-manager/org/micro-manager/acqengj/AcqEngJ/0.7.2/) in the plugins/micro-manager folder of your micromanaegr install

Great, thanks.

**Comment by dpshepherd** [comment]
> Reading through the API some more, we are starting to see a potential path forward to using the events, assuming we can synchronize everything. The ASI Tiger controller is quite picky about timing and communications when performing a constant speed stage scans.
> 
> If we place a `print(event)` call inside of the functions we call for `post_hardware_hook_fn` and `post_camera_hook_fn`, we see the following, in this order:
> 
> 1. the full list/dict due to a call to `post_hardware_hook_fn`
> 2. the full list/dict due to a call to `post_camera_hook_fn`
> 3. an empty list/dict due to a call to `post_camera_hook_fn`
> 
> I don't quite understand this. I thought these functions would just get the current event? Also, why is the camera called again at the end?

Any insight on to how to properly parse the event structure? Should the hooks be receiving all of the events? Or just the current event? If all of the events, is there another variable we can use to figure out which event we are on?

**Comment by henrypinkard** [comment]
The third one was bug that is fixed now. Each hook function will receive each event, or a list events if running as a sequence (as in your case)

**Comment by dpshepherd** [comment]
Ok, we downloaded the new AcqEng jar. The code with the demo config no longer hangs. If we define the following event structure,
```
events = []
for y in range(2):
    for z in range(10):
        evt = { 'axes': {'y': y, 'z': z}, 'y': y}
        events.append(evt)
```

And run this code ([test_dem_config.py](https://github.com/QI2lab/OPM/blob/master/Control-MM/test_demo_config.py)) using the demo config with `print(event)` in the respective hook functions, we get the following:
```
stage
{'axes': {'y': 0, 'z': 0}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 0}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 0, 'z': 1}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 1}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 0, 'z': 2}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 2}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 0, 'z': 3}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 3}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 0, 'z': 4}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 4}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 0, 'z': 5}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 5}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 0, 'z': 6}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 6}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 0, 'z': 7}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 7}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 0, 'z': 8}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 8}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 0, 'z': 9}, 'y': 0, 'properties': []}
camera
{'axes': {'y': 0, 'z': 9}, 'y': 0, 'properties': []}
stage
{'axes': {'y': 1, 'z': 0}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 0}, 'y': 1, 'properties': []}
stage
{'axes': {'y': 1, 'z': 1}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 1}, 'y': 1, 'properties': []}
stage
{'axes': {'y': 1, 'z': 2}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 2}, 'y': 1, 'properties': []}
stage
{'axes': {'y': 1, 'z': 3}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 3}, 'y': 1, 'properties': []}
stage
{'axes': {'y': 1, 'z': 4}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 4}, 'y': 1, 'properties': []}
stage
{'axes': {'y': 1, 'z': 5}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 5}, 'y': 1, 'properties': []}
stage
{'axes': {'y': 1, 'z': 6}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 6}, 'y': 1, 'properties': []}
stage
{'axes': {'y': 1, 'z': 7}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 7}, 'y': 1, 'properties': []}
stage
{'axes': {'y': 1, 'z': 8}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 8}, 'y': 1, 'properties': []}
stage
{'axes': {'y': 1, 'z': 9}, 'y': 1, 'properties': []}
camera
{'axes': {'y': 1, 'z': 9}, 'y': 1, 'properties': []}
Exception ignored in: <function JavaObjectShadow.__del__ at 0x0000024CA1870160>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
```
A few things about this,
1. If we do define the stage position, Pycromanager executes the `post_camera_hook_fun` for each event. This won't work for the stage triggered scan. We'll think about this some more, but would be great to hear your thought on how to structure the events to perform a constant speed scan at each Y without talking to the hardware (stage or camera) during the scan.
2. There is still a cleanup error occurring.
3. If we add `acq.await_completion()` after the `acq.acquire(events)` call, the Python code hangs.

**Comment by henrypinkard** [comment]
>     1. If we do define the stage position, Pycromanager executes the `post_camera_hook_fun` for each event. This won't work for the stage triggered scan. We'll think about this some more, but would be great to hear your thought on how to structure the events to perform a constant speed scan at each Y without talking to the hardware (stage or camera) during the scan.

Another small bug, try [AcqEngJ 0.7.3](https://dl.bintray.com/micro-manager/micro-manager/org/micro-manager/acqengj/AcqEngJ/0.7.3/)

**Comment by henrypinkard** [comment]
> * There is still a cleanup error occurring.
> 
> * If we add `acq.await_completion()` after the `acq.acquire(events)` call, the Python code hangs.

These might be the same thing, but I'm not able to reproduce

**Comment by dpshepherd** [comment]
> 
> Another small bug, try [AcqEngJ 0.7.3](https://dl.bintray.com/micro-manager/micro-manager/org/micro-manager/acqengj/AcqEngJ/0.7.3/)

This still occurs. If you create the following axes structure with the demo config,
```

filter_wheel = ['Chroma-D460','Chroma-HQ535','Chroma-HQ620','Chroma-HQ700']
events = []
for y in range(2):
    for z_stage in range(2):
        for c in range(len(filter_wheel)):
            for z in range(10):
                evt = { 'axes': {'x': z, 'y': y, 'z': z_stage, 'c': c}, 'y': y, 'z': z_stage, 'properties': [['Emission','Label',filter_wheel[c]]]}
                events.append(evt)
```

And run an acquisition, the camera is fired for every x axis event. What we want is 10 acquired images in a hardware sequences. If you look at the events passed in, it is a discrete event, which won't work with the constant speed stage scan.

We can achieve the desired behavior using a modified event structure, such as this one,
```
filter_wheel = ['Chroma-D460','Chroma-HQ535','Chroma-HQ620','Chroma-HQ700']
events = []
for y in range(2):
    for z_stage in range(2):
        for c in range(len(filter_wheel)):
            for z in range(10):
                if z==0:
                    evt = { 'axes': {'x': z, 'y': y, 'z': z_stage, 'c': c}, 'y': y, 'z': z_stage, 
                    'properties': [['Emission','Label',filter_wheel[c]]]}
                else:
                    evt = { 'axes': {'x': z, 'y': y, 'z': z_stage, 'c': c}}
                events.append(evt)
```

In the `post_hardware_hook_fn`, we catch the event with the stage positions and property changes. We can execute those and then remove the event from the event queue. Then those events without a stage or property are executed all in a run, which would allow for the hardware control.

This approach works for us, we just want to make sure that we are using the hook functions as intended. Is this the correct behavior?

**Comment by henrypinkard** [comment]
I'm kinda confused by the duplicate z's in your event, but I think maybe you shouldn't be adding in the z_stage to the events (outside of `axes`, that is), since you're controlling that yourself, right? You'd only need to put that in there if you want the acquisition engine to move your z stage for you.

**Comment by dpshepherd** [comment]
I agree, it is confusing because we are scanning a sample through an orthogonal light sheet. We don't scan with the Z stage, as most microscopes do. Does pycromanager assign priority to the z axis for the scan?

The setup is:
X stage -> used for constant speed stage scan.
Y stage -> used for orthogonal move to constant speed scan. 
Z stage -> used to change sample position orthogonal to both Y and Z

Y stage and Z stage cannot move or be communicated with while the X stage is scanning.

Maybe a better way of writing it is:

```
for y in range(tile_axis_positions):
    for z in range(height_axis_positions):
        for c in range(len(channel_states)):
            for x in range(scan_axis_positions):
```

**Comment by henrypinkard** [comment]
Okay I think I get it now. I just made some more tweaks, try [0.7.4](https://dl.bintray.com/micro-manager/micro-manager/org/micro-manager/acqengj/AcqEngJ/0.7.4/). Now it is sequencing over all x events and then moving czy in between each sequence. 

Also, there's a more "official" way to do channels by [putting `channels` in the acquisition event](https://pycro-manager.readthedocs.io/en/latest/apis.html#acquisition-event-specification). In theory this will help performance of file writing compared to how you do it (though it might be miniscule), and would allow you to use multichannel viewer if you ever wanted

**Comment by dpshepherd** [comment]
We will need to think about how to use `channels ` as in your documentation, because we actually use a custom DAQ controller that is in closed loop with the camera to control laser blanking and a galvo mirror. It doesn't easily fit into the standard definition of a Micromanager channel.

Because our data is skewed, we have to process it before viewing. So having official multichannel viewer support doesn't really matter to us.

We have a working script with 0.7.3: [test_events_asi.py](https://github.com/QI2lab/OPM/blob/master/Control-MM/test_events_asi.py). We will look at 0.7.4 tomorrow sometime. We still are getting the zmq.error with both the demo config and our real config, but it doesn't seem to causing a real issue.

**Comment by henrypinkard** [comment]
Probably fine to just skip the normal channel way then.

I dont think the zmq error will matter (even though it'd be nice to be able to fix it). It may cause a memory leak on the Java side (i.e. object never get garbage collected), but this is likely insignificant in size. Maybe if the speed thing doesnt get sorted out I can do a remote control and take a look at both

**Comment by dpshepherd** [comment]
Working with 0.7.4, [test_events_asi.py](https://github.com/QI2lab/OPM/blob/master/Control-MM/test_events_asi.py). We used the channels as you intended in this one for instructional purposes (ignoring our DAQ). Fixing the camera calls seems to have helped with the file writing, but we'll need to acquire a multi-terabyte dataset in one go to fully  test that.

Do you want us to translate over to a Jupyter notebook for use as an example?

Also, we have some questions on the axes order and accessing the metadata when reading the dataset as a Dask array. We'll open a separate issue for that.

**Comment by henrypinkard** [comment]
Yeah that would be great if you made it into and example

**Comment by henrypinkard** [comment]
@BinYangOptics We fixed some bugs here that explains why you were having timing issues with the acquisition hooks before. It will be available in micro-manager nightly builds after this merges (https://github.com/micro-manager/micro-manager/pull/984)

**Comment by dpshepherd** [comment]
Before we start coding, I wanted to check something about this.

If we want to hardware sequence across multiple axes instead of just the one, say our X stage and Z stage, would we just populate the event structure the same but leave off the Z stage position?

**Comment by henrypinkard** [comment]
yes if you want to control the sequencing yourself. Otherwise, if the device adapter for your z drive supports sequencing the acquisition engine will do it for you

**Comment by dpshepherd** [comment]
Great, thanks. We have to set up the sequencing on our own on the ASI controller when using the constant speed stage scans, but that just involves a few more commands to the controller before executing the acquisition.

**Comment by BinYangOptics** [comment]
@henrypinkard Thanks for the update. I will give the new version a try and see if the issues are resolved.

**Comment by dpshepherd** [comment]
Looks like I spoke too soon. Upon testing more today, we discovered the following:

1. The y and z stage positions are not being updated according to positions in the event structure. The stages move just fine in the GUI and when using the standard core functions to set positions.
2. The laser that a given channel setup flickers on, presumably when the camera starts, but then immediately turns off. The channel settings work fine in the GUI and in a MDA setup through Micromanager.

**Comment by henrypinkard** [comment]
Can you reproduce with demo config?

**Comment by dpshepherd** [comment]
Stages - yes.
Channels - no, because there is no way to diagnose if a channel stays active using the demo config.

**Comment by henrypinkard** [comment]
[0.7.5](https://dl.bintray.com/micro-manager/micro-manager/org/micro-manager/acqengj/AcqEngJ/0.7.5/) hopefully fixes shutter issue. 

Can you post the demo config code showing the problem for stages?

**Comment by dpshepherd** [comment]
[test_demo_config.py](https://github.com/QI2lab/OPM/blob/master/Control-MM/test_demo_config.py)

How to reproduce:
1. Open MM with demo config
2. Open stage position controller so that you can see the stage positions. Make sure 'polling updates' is on.
3. Run python code above
4. X,Y,Z stages will make the initial move, which comes from calls to the core functions on lines 57-66
5. Y,Z stages do not move during Acquisition

**Comment by henrypinkard** [comment]
Okay, I think its now fixed with [0.7.6](https://dl.bintray.com/micro-manager/micro-manager/org/micro-manager/acqengj/AcqEngJ/0.7.5/), but I did have to make some modifications to your script: 1) you have to define both x and y positions since the core API only has calls to set both. 2) You've redundantly specified `channel` and `c` axis. Pick one or you'll have duplicate axes where you have to specify position to both

**Comment by dpshepherd** [comment]
1) I don't quite understand this, can you post your example? How does it know to sequence on the x axis if you are specifying the position?
2) that's a mistake on our part

**Comment by henrypinkard** [comment]
```
    for y in range(2):
        for z in range(2):
            for c in range(len(channel_configs)):
                for x in range(10):
                    evt = {'axes': {'x': x, 'y': y, 'z': z, 'c': c}, 'x': 400, 'y': y * 1000, 'z': z * 100,
                           # 'channel': {'group': 'Channel', 'config': channel_configs[c]}
                           }
                    events.append(evt)
```

Sequence can be run over anything in which position doesn't change or it does change but the hardware supports sequencing. 

The api doesnt consider X and Y separately, so if you ever specify a position for one you need to add the other.

**Comment by dpshepherd** [comment]
Ah, I understand now. We will both this and the shutter fix a go and report back.

**Comment by dpshepherd** [comment]
The stages are now working. Channel selection is not.

The same channels we are using for pycro-manager do work if we run a MDA through the MM GUI.

**Comment by henrypinkard** [comment]
Fixed this morning (https://github.com/micro-manager/AcqEngJ/pull/27). Just pushed to[ 0.7.7](https://dl.bintray.com/micro-manager/micro-manager/org/micro-manager/acqengj/AcqEngJ/0.7.7/)

**Comment by dpshepherd** [comment]
AcqEng 0.7.7 causes the following new error,
```
Exception ignored in: <function JavaObjectShadow.__del__ at 0x00000291C61231F0>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
```

and the scan does not start. AcqEnJ 0.7.6 ran the scan, but did not correctly set the channels. This is still pycromanager 0.5.0. Do I need to pull from source and rebuild?

**Comment by henrypinkard** [comment]
Thats weird those things shouldnt be related at all. Maybe try restarting reinitializing everything?

**Comment by dpshepherd** [comment]
Rebooted the setup and got this error with our hardware,
```
Exception ignored in: <function JavaObjectShadow.__del__ at 0x00000295745731F0>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
Exception ignored in: <function JavaObjectShadow.__del__ at 0x00000295745731F0>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
```

and this error with the demo config,
```
Exception ignored in: <function JavaObjectShadow.__del__ at 0x00000284FF7C11F0>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
```

**Comment by dpshepherd** [comment]
Tried a fresh install of MM 2.0 and downloading AcqEnJ 0.7.7 again. Still get this,
```
Exception ignored in: <function JavaObjectShadow.__del__ at 0x00000184DA5A31F0>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpsheph1\Anaconda3\envs\pycromanager\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq/backend/cython/socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq/backend/cython/socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
```

It appears to happen after the call to `post_camera_hook_fn` in both the real config and demo config. The NDviewer does pop up, but no channel information populates it. It also says "(Finished)" at the top of the NDviewer window next to the dataset name.

**Comment by henrypinkard** [comment]
CAn you post your demo code, seems to be working fine with what you have me before

**Comment by dpshepherd** [comment]
It's the same code. Verified on my laptop as well with fresh install of Micromanager and pycromanager.

**Comment by dpshepherd** [comment]
Maybe a package or python version mismatch? Here is the conda list for mine,
```
# Name                    Version                   Build  Channel
ca-certificates           2020.6.20            hecda079_0    conda-forge
certifi                   2020.6.20        py38h9bdc248_2    conda-forge
dask                      2.30.0                   pypi_0    pypi
numpy                     1.19.2                   pypi_0    pypi
openssl                   1.1.1h               he774522_0    conda-forge
pip                       20.2.4                     py_0    conda-forge
pycromanager              0.5.0                    pypi_0    pypi
python                    3.8.6           h60c2a47_0_cpython    conda-forge
python_abi                3.8                      1_cp38    conda-forge
pyyaml                    5.3.1                    pypi_0    pypi
pyzmq                     19.0.2                   pypi_0    pypi
setuptools                49.6.0           py38h9bdc248_2    conda-forge
sqlite                    3.33.0               he774522_1    conda-forge
toolz                     0.11.1                   pypi_0    pypi
vc                        14.1                 h869be7e_1    conda-forge
vs2015_runtime            14.16.27012          h30e32a0_2    conda-forge
wheel                     0.35.1             pyh9f0ad1d_0    conda-forge
wincertstore              0.2             py38h32f6830_1005    conda-forge
zmq                       0.0.0                    pypi_0    pypi
```

**Comment by henrypinkard** [comment]
I have pyzmq 18.1.1. Thats the only difference I see

Can you repost the demo code you have just in case im missing something?

**Comment by dpshepherd** [comment]
Ok, here it is: [test_demo_config.py](https://github.com/QI2lab/OPM/blob/master/Control-MM/test_demo_config.py)

**Comment by dpshepherd** [comment]
Ok, downgrading to pyzmq 18.1.1 and re-running gives the following error,
```
Exception ignored in: <function JavaObjectShadow.__del__ at 0x0000025716064F70>
Traceback (most recent call last):
  File "C:\Users\dpshe\miniconda3\envs\test_bed_pycro\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpshe\miniconda3\envs\test_bed_pycro\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpshe\miniconda3\envs\test_bed_pycro\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpshe\miniconda3\envs\test_bed_pycro\lib\site-packages\zmq\sugar\socket.py", line 400, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq\backend\cython\socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: not a socket
```

**Comment by henrypinkard** [comment]
I don't think this is the error youre seeing, but I fixed another thing in the acq engine. Try [0.7.8](https://dl.bintray.com/micro-manager/micro-manager/org/micro-manager/acqengj/AcqEngJ/0.7.8/)

**Comment by dpshepherd** [comment]
The demo config and demo code now work. I still get an error after the scan has finished, both on my laptop and instrument computer.

We can test on the hardware tomorrow.

```
Exception ignored in: <function JavaObjectShadow.__del__ at 0x000001B7185C70D0>
Traceback (most recent call last):
  File "C:\Users\dpshe\miniconda3\envs\test_bed_pycro\lib\site-packages\pycromanager\core.py", line 341, in __del__
    self._close()
  File "C:\Users\dpshe\miniconda3\envs\test_bed_pycro\lib\site-packages\pycromanager\core.py", line 331, in _close
    self._socket.send(message)
  File "C:\Users\dpshe\miniconda3\envs\test_bed_pycro\lib\site-packages\pycromanager\core.py", line 72, in send
    self._socket.send(bytes(json.dumps(message), 'utf-8'))
  File "C:\Users\dpshe\miniconda3\envs\test_bed_pycro\lib\site-packages\zmq\sugar\socket.py", line 416, in send
    return super(Socket, self).send(data, flags=flags, copy=copy, track=track)
  File "zmq\backend\cython\socket.pyx", line 728, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 769, in zmq.backend.cython.socket.Socket.send
  File "zmq\backend\cython\socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: not a socket
```

**Comment by henrypinkard** [comment]
I just pushed some logging changes (https://github.com/micro-manager/pycro-manager/pull/166) to try to figure out what this error is. Can you try running from python source and see what it prints?

**Comment by dpshepherd** [comment]
The actual hardware appears to be working with 0.7.8. We can stress test this weekend.

I ran from your pull request and get the following for the demo code,
```
Unknown error
<pycromanager.core.JavaSocket object at 0x0000024786DE1C40> <class 'pycromanager.core.JavaSocket'>
Exception ignored in: <function JavaObjectShadow.__del__ at 0x0000024786136040>
Traceback (most recent call last):
  File "C:\Users\dpsheph1\Anaconda3\envs\pycro_test\lib\site-packages\pycromanager-0.5.0-py3.8.egg\pycromanager\core.py", line 345, in __del__
  File "C:\Users\dpsheph1\Anaconda3\envs\pycro_test\lib\site-packages\pycromanager-0.5.0-py3.8.egg\pycromanager\core.py", line 336, in _close
  File "C:\Users\dpsheph1\Anaconda3\envs\pycro_test\lib\site-packages\pycromanager-0.5.0-py3.8.egg\pycromanager\core.py", line 85, in receive
  File "zmq\backend\cython\socket.pyx", line 791, in zmq.backend.cython.socket.Socket.recv
  File "zmq\backend\cython\socket.pyx", line 824, in zmq.backend.cython.socket.Socket.recv
  File "zmq\backend\cython\socket.pyx", line 135, in zmq.backend.cython.socket._check_closed
zmq.error.ZMQError: Unknown error
```

**Comment by henrypinkard** [comment]
Hmm still not sure, but one other thing I just thought of--If you make youre `Acquisition` with `debug=True` as an argument, you it will print every command into the micromanager corelog. Then I might be able to figure out what call is triggering the error

**Comment by dpshepherd** [comment]
I tried to upload here, but got an error. Sent via email.

**Comment by henrypinkard** [comment]
I guess I was wrong, it only prints on the java side when you're running from source. 

I added same thing to python side (https://github.com/micro-manager/pycro-manager/pull/166), could you pull latest python source, try again, and send me what it prints?

**Comment by dpshepherd** [comment]
Pulled and emailed.

**Comment by henrypinkard** [comment]
I was able to reproduce on windows. It only seems to occur in very specific circumstances (windows, 2 hooks, both getting core). Wasn't able to figure out why (maybe a zmq error?) but it doesnt seem to cause any problems in your setup. Made a new issue to track it (https://github.com/micro-manager/pycro-manager/issues/169)

Also made a PR to add in the changes to AcqEngJ that fix the original issue here
https://github.com/micro-manager/micro-manager/pull/991
